src/compile/compile.ts

```diff
@@ -65,23 +69,43 @@ function getTopLevelProperties(topLevelSpec: TopLevel<any>, config: Config) {
  * Note: this couldn't be `model.assemble()` since the top-level model
  * needs some special treatment to generate top-level properties.
  */
-function assembleTopLevelModel(model: Model, topLevelProperties: TopLevelProperties) {
+function assembleTopLevelModel(model: Model, topLevelProperties: TopLevelProperties & LayoutSizeMixins) {
   // TODO: change type to become VgSpec
 
   // Config with Vega-Lite only config removed.
   const vgConfig = model.config ? stripAndRedirectConfig(model.config) : undefined;
 
   // autoResize has to be put under autosize
-  const {autoResize, ...topLevelProps} = topLevelProperties;
+  const {autosize, ...topLevelProps} = topLevelProperties;
 
   const title = model.assembleTitle();
   const style = model.assembleGroupStyle();
 
+  let layoutSignals = model.assembleLayoutSignals();
+
+  // move width and height signals with values to top level
+  layoutSignals = layoutSignals.filter(signal => {
+    if ((signal.name === 'width' || signal.name === 'height') && signal.value !== undefined) {
+      topLevelProps[signal.name] = signal.value;
+      return false;
+    }
+    return true;
+  });
+
+  if (autosize === 'fit' || (autosize && isAutosize(autosize) && autosize.type === 'fit')) {
+    if (!topLevelProps.width || !topLevelProps.height) {
+      log.warn(log.message.FIT_DATA_DRIVEN);
+    }
+
+    if (!(model instanceof UnitModel) && !(model instanceof LayerModel)) {
+      log.warn(log.message.FIT_NON_SINGLE);
```

Comment:
Using this fit thing with facet makes the visualization output nothing.  Won't it be better to drop it? 

In fact, if possible, we may want to make the autosize fit only available for unit / layer in the schema?

Reply:
I could do it in the schema but that requires changing all instances of TopLevelProperties and TopLevel and probably more. I'll drop it in the code.

Reply:
> changing all instances of TopLevelProperties and TopLevel and probably more

I doubt if it is really that much work.

Reply:
I'd have to make TopLevelProperties a generic. Seems overkill in this case.